packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -247,10 +248,13 @@ class FlutterCommandRunner extends CommandRunner<void> {
         final bool redirectedCompletion = !globals.stdio.hasTerminal &&
             (topLevelResults.command?.name ?? '').endsWith('-completion');
         final bool isMachine = machineFlag || ci || redirectedCompletion;
+        final bool explicitVersionCheckDisabledEnvironment =
+            globals.platform.environment['FLUTTER_VERSION_CHECK'] == 'false';
         final bool versionCheckFlag = topLevelResults['version-check'] as bool? ?? false;
         final bool explicitVersionCheckPassed = topLevelResults.wasParsed('version-check') && versionCheckFlag;
 
         if (topLevelResults.command?.name != 'upgrade' &&
+            !explicitVersionCheckDisabledEnvironment &&
```

Comment:
the logic in this if block is very confusing. how about if we pass all of these boolean values to `globals.flutterVersion.checkFlutterVersionFreshness()` and then have that method determine if it should actually run or not? Then it would be easier to unit test this logic.

Also, this change will need tests to be landed.